Source/Engine/World.swift

```diff
@@ -38,7 +38,15 @@ public extension World {
 
     mutating func update(timeStep: Double, input: Input) {
         player.direction = player.direction.rotated(by: input.rotation)
-        player.velocity = player.direction * input.speed * player.speed
+        let cosine = player.direction.cosine(to: player.initialDirection)
+        let sine = (1 - cosine * cosine).squareRoot()
+        let rotation: Rotation
+        if player.direction.y <= 0 {
+            rotation = Rotation(sine: -sine, cosine: cosine)
+        } else {
+            rotation = Rotation(sine: sine, cosine: cosine)
+        }
+        player.velocity = input.speed.rotated(by: rotation) * player.speed
         player.position += player.velocity * timeStep
```

Comment:
As I suspected, this can actually be simplified to just the following.

```swift
    mutating func update(timeStep: Double, input: Input) {
        player.direction = player.direction.rotated(by: input.rotation)
        let forwardVelocity = player.direction * input.speed.x * player.speed
        let strafingVelocity = player.direction.orthogonal * input.speed.y * player.speed
        player.velocity = forwardVelocity + strafingVelocity
        player.position += player.velocity * timeStep
        while let intersection = player.intersection(with: map) {
            player.position -= intersection
        }
    }
```

The fact that I had to use `input.speed.x` for forward motion and `input.speed.y` for strafing confused me a little, but I realized that it's because you're passing `leftInputVector.orthogonal` as the input speed in `ViewController`.